### PR TITLE
Deprecating change directory

### PR DESCRIPTION
**Pull request by herbelin**
In case there is a wish to eventually remove `Cd` (which does not make sense is a relocatable document).

Closes #16119.

Depends on #17392 (merged) and #16126 (merged).

- [x] Added **changelog**.

### CODE REVIEW

**Comment by jfehrle** [comment]
This has a lot more than just a deprecation, which is just the last commit.  Do you expect the others will be committed before this one in other PRs?

File: lib/system.ml [hunk 1]
```diff
@@ -56,6 +56,13 @@ let exists_dir dir =
   let dir = if Sys.os_type = "Win32" then strip_trailing_slash dir else dir in
   try Sys.is_directory dir with Sys_error _ -> false
 
+let rec mkdirhier dir =
```
Comment (Alizter) [inline lib/system.ml:59]: I think this should just be called `mkdir`.

Comment (Alizter) [reply lib/system.ml:59]: Add that it creates the entire subdirectory in the mli.

Comment (herbelin) [reply lib/system.ml:59]: I used mkdirhier in reference to the Unix command of same name (but I agree that mkdir is simpler).

**Comment by herbelin** [comment]
> Do you expect the others will be committed before this one in other PRs?

Yes, the other ones justify that we may eventually remove `Cd`.

> The doc for Cd should mention the deprecation. Also needs a changelog.

Yes, it has to be done. I will do eventually, when the decisions taken about the other PRs are clearer.

File: doc/sphinx/addendum/extraction.rst [hunk 2]
```diff
@@ -103,6 +103,11 @@ in the Coq sources.
 
 .. cmd:: Cd {? @string }
 
+   .. deprecated:: 8.20
+
+      Use instead the command line option :n:`-output-directory` (see
+      :ref:`command-line-options`).
```
Comment (jfehrle) [inline doc/sphinx/addendum/extraction.rst:109]:
```suggestion
      Use the command line option :n:`-output-directory` instead (see
      :ref:`command-line-options`).
```

**Comment by herbelin** [comment]
Changelog added

File: doc/changelog/08-vernac-commands-and-options/17403-master+deprecating-change-directory.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+- **Deprecated:**
+  Command :cmd:`Cd` is deprecated, use instead the command line option
+  `-output-directory` (see :ref:`command-line-options`) or, for
+  extraction, :flag:`Set Extraction Output Directory`
```
Comment (jfehrle) [inline doc/changelog/08-vernac-commands-and-options/17403-master+deprecating-change-directory.rst:4]:
```suggestion
  The :cmd:`Cd` command is deprecated, instead use the command line option
  `-output-directory` (see :ref:`command-line-options`) or, for
  extraction, :flag:`Set Extraction Output Directory`
```

Comment (herbelin) [reply doc/changelog/08-vernac-commands-and-options/17403-master+deprecating-change-directory.rst]: Good to me. Thanks.

File: doc/sphinx/addendum/extraction.rst [hunk 4]
```diff
@@ -103,6 +103,12 @@ in the Coq sources.
 
 .. cmd:: Cd {? @string }
 
+   .. deprecated:: 8.20
+
+      Use instead the command line option :n:`-output-directory` (see
```
Comment (jfehrle) [inline doc/sphinx/addendum/extraction.rst:108]:
```suggestion
      Instead use the command line option :n:`-output-directory` (see
```

File: vernac/synterp.ml [hunk 5]
```diff
@@ -318,9 +318,14 @@ let synterp_declare_ml_module ~local l =
   let l = List.map expand l in
   Mltop.declare_ml_modules local l
 
+let warn_chdir = CWarnings.create ~name:"change-dir-deprecated" ~category:Deprecation.Version.v8_18
+    (fun () -> strbrk "Command \"Cd\" is deprecated." ++ spc () ++
+               strbrk "Use command-line \"-output-directory dir\" instead.")
```
Comment (jfehrle) [inline vernac/synterp.ml:323]: maybe also mention Extraction Output Directory?

Comment (herbelin) [reply vernac/synterp.ml]: Done

File: doc/sphinx/addendum/extraction.rst (hunk 4 above)
Comment (herbelin) [reply doc/sphinx/addendum/extraction.rst]: Done

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by SkySkimmer — APPROVED** [review]
We may also want to change ppvernac to produce `Pwd` instead of `Cd` from `VernacChdir None` and maybe also warn on argument-less Cd (but not Pwd?)
But this doesn't need to be done in this PR.

**Comment by herbelin** [comment]
>  We may also want to change ppvernac to produce Pwd instead of Cd from VernacChdir None and maybe also warn on argument-less Cd (but not Pwd?)

I added a commit for it (but ok also to have this commit in an other PR).

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**